kafka-client/src/main/java/dev/responsive/kafka/internal/clients/OriginEventRecorder.java

```diff
@@ -0,0 +1,210 @@
+import dev.responsive.kafka.internal.license.server.model.OriginEventsReportRequestV1;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.function.BiConsumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.streams.StreamsConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class OriginEventRecorder implements ResponsiveConsumer.Listener {
+
+  private static final Logger LOG = LoggerFactory.getLogger(OriginEventRecorder.class);
+
+  static final String ORIGIN_EVENT_HEADER_KEY = "_roe";
+  static final byte[] ORIGIN_EVENT_MARK = new byte[] {0x1};
+
+  private final Duration reportInterval;
+  private final BiConsumer<Map<TopicPartition, Long>, String> reportOriginEvents;
+  private final String threadId;
+
+  @GuardedBy("this")
+  private final Map<TopicPartition, Long> originEvents = new HashMap<>();
+  @GuardedBy("this")
+  private long lastCommitTs = 0L;
+
+  public OriginEventRecorder(
+      final String threadId,
+      final ResponsiveConfig config,
+      LicenseInfo license
+  ) {
+    this(
+        threadId,
+        new Reporter(config, license),
+        config.getLong(ResponsiveConfig.ORIGIN_EVENT_REPORT_INTERVAL_MS_CONFIG)
+    );
+  }
+
+  @VisibleForTesting
+  OriginEventRecorder(
+      final String threadId,
+      final BiConsumer<Map<TopicPartition, Long>, String> reportOriginEvents
+  ) {
+    this(threadId, reportOriginEvents, 10_000);
+  }
+
+  private OriginEventRecorder(
+      final String threadId,
+      final BiConsumer<Map<TopicPartition, Long>, String> reportOriginEvents,
+      final long reportIntervalMs
+  ) {
+    this.threadId = threadId;
+    this.reportOriginEvents = reportOriginEvents;
+    this.reportInterval = Duration.ofMillis(reportIntervalMs);
+  }
+
+  private synchronized void inc(final TopicPartition tp) {
+    originEvents.compute(tp, (k, v) -> v == null ? 1 : v + 1);
+  }
+
+  @Override
+  public <K, V> ConsumerRecords<K, V> onPoll(final ConsumerRecords<K, V> records) {
+    for (final ConsumerRecord<K, V> record : records) {
+      final var header = record.headers().lastHeader(ORIGIN_EVENT_HEADER_KEY);
+      if (header == null) {
+        record.headers().add(ORIGIN_EVENT_HEADER_KEY, ORIGIN_EVENT_MARK);
+        inc(new TopicPartition(record.topic(), record.partition()));
+      }
+    }
+    return records;
+  }
+
+  @Override
+  public void onCommit(final Map<TopicPartition, OffsetAndMetadata> offsets) {
+    final var now = System.currentTimeMillis();
```

Comment:
The offsets being committed aren't necessarily what's been polled so far. Streams puts the polled records into its own buffer, and can commit at any time, even if the internal buffer isn't drained. So technically we should only be reporting origin events up to the offsets being committed. How to do that is tricky. Need to think about it some more.

Reply:
OK I fixed this and I believe the previous one as well by tracking origin events in a bitset instead of just as a long, and then counting that bitset on commit. It's a bit heavyweight, but I think it should be fine since if we are tracking 100K offsets per commit we'll only need ~1.5k longs which should be hardly any memory overhead and 1.5k bitwise ops for counting the number of 1s shouldn't take much either. PTAL

Reply:
yeah this is a good catch, what's been polled =/= what's being committed. I think we could find other ways to work around this (for example by wrapping the `ConsumerRecord` class) but your bitwise solution seems efficient and simple enough.

Just please explain why it does what it does in the javadocs because without this context it's pretty hard to understand what it's there for lol